### PR TITLE
ActiveStorage reflection

### PR DESCRIPTION
**Pull request by kddnewton**
Add the ability to query for the attachments that have been defined on a model through the `ActiveRecord::Base::defined_attachments` method. 

Useful for writing things on top of ActiveStorage that can loop through models and their attachments to perform things in bulk.

### CODE REVIEW

**Comment by rafaelfranca** [comment]
I'm fine with a reflection API but we should follow the same interface and similar implementation of the association and aggregation reflection API. https://github.com/rails/rails/blob/375a4143cf5caeb6159b338be824903edfd62836/activerecord/lib/active_record/reflection.rb

**Comment by kddeisz** [comment]
@rafaelfranca I updated it to hook into `ActiveRecord::Reflection`, let me know what you think

File: activerecord/lib/active_record/reflection.rb [hunk 1]
```diff
@@ -1038,5 +1039,7 @@ def aliased_table
 
       def all_includes; yield; end
     end
+
+    ActiveSupport.run_load_hooks(:active_record_reflection, self)
```
Comment (rafaelfranca) [inline activerecord/lib/active_record/reflection.rb]: I think we can do this without adding a new load hook to the public API. The `class_attribute` should be added to the `ActiveRecord::Base` class. The `ReflectionExtensions` can be added directly to `ActiveRecord::Reflection` without need a `on_load` hook.

File: activestorage/lib/active_storage/reflection.rb [hunk 2]
```diff
@@ -0,0 +1,69 @@
+# frozen_string_literal: true
+
+module ActiveStorage
+  module Reflection
+    # Holds all the metadata about a has_one_attached attachment as it was
+    # specified in the Active Record class.
+    class HasOneAttachedReflection <
+        ::ActiveRecord::Reflection::MacroReflection #:nodoc:
```
Comment (rafaelfranca) [inline activestorage/lib/active_storage/reflection.rb]: Why you need the `::` here? Also keep everything in the same line.

File: activerecord/lib/active_record/reflection.rb (hunk 1 above)
Comment (kddeisz) [reply activerecord/lib/active_record/reflection.rb]: Sure that makes sense. I was just trying to keep the ActiveStorage stuff to itself, but happy to move it over.

File: activestorage/lib/active_storage/reflection.rb (hunk 2 above)
Comment (kddeisz) [reply activestorage/lib/active_storage/reflection.rb]: Top level constant just for the milliseconds it would shave off the constant lookup - I'll remove

**Review by rafaelfranca — APPROVED** [review]
I ok with this current implementation. Even that those reflections can only be added once you have active storage loaded I think it is fine to have the reflections always on Active Record.

I'd prefer to have it totally separated though, and it is possible to do without introducing new `on_load` hooks.

I'll check if other member of the team before implementing it.

@georgeclaghorn @matthewd options?

**Comment by georgeclaghorn** [comment]
I’m okay with this implementation, too, but I’d prefer for it to live in Active Storage.

**Comment by kddeisz** [comment]
I'd be happy to refactor, just wasn't sure how to hook it in up the engine to ensure `ActiveRecord::Reflection` had been loaded without adding an `on_load` hook to `ActiveSupport`. Any ideas? Honestly I can probably just put it into the `on_load` for `:active_record`, I'll bet that would work.

**Comment by kddeisz** [comment]
Okay so refactored and this works. I can move these modules into ActiveRecord or leave them here, let me know which is preferable.

File: activestorage/lib/active_storage/reflection.rb [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    # Holds all the metadata about a has_many_attached attachment as it was
+    # specified in the Active Record class.
+    class HasManyAttachedReflection < ActiveRecord::Reflection::MacroReflection #:nodoc:
+      def macro
+        :has_many_attached
+      end
+    end
+
+    module ReflectionExtension
```
Comment (rafaelfranca) [inline activestorage/lib/active_storage/reflection.rb]: I think this one should be `:nodoc:` like all methods in Reflection are.

File: activestorage/lib/active_storage/reflection.rb [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+      def reflection_class_for(macro)
```
Comment (rafaelfranca) [inline activestorage/lib/active_storage/reflection.rb]: This method should be private

**Review by rafaelfranca — COMMENTED** [review]
Yeah, now it is better

**Comment by kddeisz** [comment]
@rafaelfranca updated!

**Pull request merged**